### PR TITLE
Schematic decorations: Add 'place_offset_y' placement parameter

### PR DESCRIPTION
**Pull request by paramat**
For precise control of schematic vertical position relative to the
'place_on' node.
Avoids workarounds that add empty nodes to a schematic and therefore
reduce performance.

Also remove long-unused decoration cutoff code.
/////////////

screenshot_20170906_225408

Requested in #6375 and useful anyway.

Shifts the base of the schematic vertically by a defined amount. Default 0.
A more precise vertical shift than the flag 'deco place center y' which can only shift the schematic downwards by half the schematic height. This flag is still usable for mod compatibility and overrides 'y offset', if the flag is not used then 'y offset' is used.
Value +1 moves the base of a schematic to being above and resting on ground level.
Also add a check that the schematic base has not moved below the voxelmanip.

Also remove long-unused 'deco cutoff' stuff.

Until now any precise vertical shifting of a schematic involves adding often large amounts of empty nodes to the schematic, slowing mapgen.

### CODE REVIEW

**Comment by paramat** [comment]
Tested, just need to add docs.

**Comment by paramat** [comment]
Docs added, now ready for review.

File: doc/lua_api.txt [hunk 1]
```diff
@@ -4743,6 +4743,12 @@ The Biome API is still in an experimental phase and subject to change.
     --  ^ Flags for schematic decorations.  See 'Schematic attributes'.
         rotation = "90" -- rotate schematic 90 degrees on placement
     --  ^ Rotation can be "0", "90", "180", "270", or "random".
+        y_offset = 1,
```
Comment (Wuzzy2) [inline doc/lua_api.txt]: Wait. If the default offset is 1, it means the schematic is above 1 node relative to the place_node. This will break all existing schematics by moving them up 1 node. So shouldn't the default offset be 0?

File: doc/lua_api.txt [hunk 2]
```diff
@@ -4743,6 +4743,12 @@ The Biome API is still in an experimental phase and subject to change.
...
+    --  ^ Y offset of the base node layer of the schematic relative to the
+    --  ^ 'place_on' node, can be positive or negative.
+    --  ^ If the flag 'place_center_y' is set this parameter is ignored.
+    --  ^ If absent or zero no vertical offset is applied and the schematic
+    --  ^ base node layer is at the level of the 'place_on' node.
```
Comment (Wuzzy2) [inline doc/lua_api.txt]: Explain what this means practically. I.e. that the schematic is “buried into” the place_on node by default.

File: doc/lua_api.txt (hunk 1 above)
Comment (paramat) [reply doc/lua_api.txt]: The default is zero, 1 is used here as an example, because other parameters are documented with example-type values.

File: doc/lua_api.txt (hunk 2 above)
Comment (paramat) [reply doc/lua_api.txt]: It is clear and precise language already, i'm not going to dumb it down :)

File: doc/lua_api.txt (hunk 1 above)
Comment (paramat) [reply doc/lua_api.txt]: But, i should make the default value clearer.

**Comment by paramat** [comment]
Updated to be clearer. Since 0 is the most likely value i have changed the example to 0.

File: doc/lua_api.txt (hunk 1 above)
Comment (Wuzzy2) [reply doc/lua_api.txt]: Oops. I thought the “parameter = value” syntax always showed the default value. At least this seems to be the case with most things in `lua_api.txt`.

**Comment by Wuzzy2** [comment]
I made a quick test. Works nice! This is exactly what I needed. :-)

The fact that you can offset freely is very interesting, now we just need a creative idea. :D

File: src/mg_decoration.cpp [hunk 3]
```diff
@@ -336,11 +277,17 @@ size_t DecoSchematic::generate(MMVManip *vm, PcgRandom *pr, v3s16 p)
...
-		p.Y -= (schematic->size.Y - 1) / 2;
 	if (flags & DECO_PLACE_CENTER_Z)
 		p.Z -= (schematic->size.Z - 1) / 2;
 
+	if (flags & DECO_PLACE_CENTER_Y)
+		p.Y -= (schematic->size.Y - 1) / 2;
+	else
+		p.Y += y_offset;
+	// Check shifted schematic base is in voxelmanip
+	if (p.Y < vm->m_area.MinEdge.Y)
```
Comment (SmallJoker) [inline src/mg_decoration.cpp]: How about `MaxEdge.Y`? Do we also need checks for the upper limits?

Comment (paramat) [reply src/mg_decoration.cpp]: See line 196 in placeDeco(), already done.

**Comment by paramat** [comment]
Adding root systems to trees is one idea.
But also this means we can remove the empty nodes from MTGame bushes, logs and waterlily schematics that are added as workarounds.

**Comment by paramat** [comment]
LINT complains of unrelated file:
```
The file src/network/connectionthreads.h is not compliant with the coding style:
--- src/network/connectionthreads.h	2017-09-07 01:23:29.325945649 +0000
+++ /dev/fd/63	2017-09-07 01:23:42.851764580 +0000
@@ -52,8 +52,8 @@
 private:
 	void runTimeouts(float dtime);
 	void rawSend(const BufferedPacket &packet);
-	bool rawSendAsPacket(u16 peer_id, u8 channelnum, SharedBuffer<u8> data,
-			bool reliable);
+	bool rawSendAsPacket(
+			u16 peer_id, u8 channelnum, SharedBuffer<u8> data, bool reliable);
 
 	void processReliableCommand(ConnectionCommand &c);
 	void processNonReliableCommand(ConnectionCommand &c);
```

**Comment by paramat** [comment]
51002b1629c69adb0053bc36d3667d2b790e5c21
